src/Compilers/Core/Portable/FileSystem/PathUtilities.cs

```diff
@@ -773,6 +775,46 @@ public static bool IsValidFilePath([NotNullWhen(true)] string? fullPath)
         public static string NormalizeWithForwardSlash(string p)
             => DirectorySeparatorChar == '/' ? p : p.Replace(DirectorySeparatorChar, '/');
 
+        /// <summary>
+        /// Takes a normalized full path from <see cref="NormalizeWithForwardSlash(string)"/> and expands any '..' or '.'
+        /// into their equivalent representation.
+        /// </summary>
+        /// <returns>An equivalent path that does not contain any '..' or '.' path parts.</returns>
+        public static string ExpandNormalizedFullPathWithRelativeParts(string p)
+        {
+            if (!IsAbsolute(p))
+            {
+                // if this isn't an absolute path we can't expand it correctly without a root path
+                return p;
+            }
+
+            // GetPathParts also removes any instances of '.'
+            var parts = GetPathParts(p);
+
+            // On windows we need to skip the volume specifier, but remember it for re-joining later
+            var volumeSpecifier = (!IsUnixLikePlatform) ? string.Empty : p.Substring(0, 2);
+
+            // Skip the root directory
+            var toSkip = IsUnixLikePlatform ? 1 : 2;
+            Debug.Assert(parts[toSkip - 1] == string.Empty);
+
+            var resolvedParts = ArrayBuilder<string>.GetInstance();
+            foreach (var part in parts.Skip(toSkip))
+            {
+                if (!part.Equals(ParentRelativeDirectory))
+                {
+                    resolvedParts.Push(part);
+                }
+                // /../../file is considered equal to /file, so we only process the parent relative directory info if there is actually a parent
+                else if (resolvedParts.Count > 0)
+                {
+                    resolvedParts.Pop();
+                }
+            }
+
+            return volumeSpecifier + '/' + string.Join("/", resolvedParts.ToArrayAndFree());
```

Comment:
❓ `resolvedParts.ToArrayAndFree()` unsure if its cheaper to enumerate the builder, then free it, rather than copying to an array first. (Although I imagine its probably not a massive difference either way)

Reply:
I'd suggest passing in the `ArrayBuilder<string>` as is, then freeing after the call, to avoid the array allocation.